### PR TITLE
Refactor: implement tag-based deployment

### LINKED ISSUE #1527: Consider a "Deploy on Release" model
Today, our [release process](http://docs.calitp.org/benefits/deployment/release/) requires a number of manual steps, and at least 1 Pull Request requiring interaction from at least 2 people (the author and an approver). This works well but is cumbersome, and could be more efficient especially for getting urgent or timely changes out.

Further, we have 3 separate "versions" of the code across the `dev`, `test`, and `prod` branches. The explicit branch naming makes it easy to see where each environment is, but it also means we're constantly merging and needing to remember the state of each branch (e.g. is `test` ready to release to `prod`? Did we merge that fix into `test` yet?)

The final step in our release process today is to tag and create a [GitHub release](https://github.com/cal-itp/benefits/releases). 

I propose we invert this model and instead make the tagging and GitHub release the initial step that kicks off the rest of the deployment.

### Branches

* We remove the 3 environment branches: `dev`, `test`, and `prod`
* All new code is merged to a single default branch `main`
* `dev` would essentially be renamed to `main` but otherwise function the same

### Deployments

* Merging to `main` deploys to the `dev` environment (just like merging to `dev` today)
* Creating a GitHub pre-release deploys to the `test` environment
* Create a GitHub release deploys to the `prod` environment

## Discussion

Creating a GitHub Release requires an associated `git tag`. 

Today we tag the `prod` branch at the end of the process, just before creating the GitHub Release.

Under this new proposed model:

* All the code (generally) is in a single branch / timeline `main`
* We tag at the beginning of the process to signal a release
* We have far more tags on the `main` branch, as both `test` and `prod` releases are tagged
* Releases are just specific points in the timeline ("forward" or "backward")
* We could think about using tag names or partial names to signify specific features (e.g. `sacrt` or `veterans`)
* `release == tag == commit`

### Pros

* Simpler release process: choose a commit, tag it, push the tag, make a (pre) release in GitHub
* Reduced friction to release: no more PRs for approval, drastically reduced [release checklist](https://github.com/cal-itp/benefits/issues/new?assignees=&labels=release&projects=&template=release.yml&title=Make+a+Release)
* Easier to get features and fixes out sooner
* Features and fixes can be released to `test` / `prod` independent of where or whether the commit is in `main`

### Cons

* There will be a lot of tags
* It may be _too easy_ to make a deployment? E.g. broken code, tests not passing
* Tagging/releasing from multiple branches, forward and backward in the timeline, may be even more overhead than our 3 straightforward branches

### Open questions

* Are existing branch protections for who can write to `main` enough? Look at the beta [tag protections](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules)?
  * Who can add tags and make releases (GitHub permissions)?
* Do we want to support deployments from other branches besides `main`?
* How will we handle updating the app's version? Can't update a file manually anymore, get it from git tags/sha?
* How can we ensure our checks have run on the commit to be deployed? e.g. `pytest`, `cypress`, `pre-commit`, `codeql` etc.

## Additional context

This model is typical of how we deploy Python packages, e.g. [`eligibility-api`](https://github.com/cal-itp/eligibility-api/blob/main/.github/workflows/publish.yml).

This model gets us closer to going full CI/CD #1253

### PR DESCRIPTION
**Pull request by lalver1**
This PR implements the second step of code changes needed to transition the benefits repository to tag-based deployments as described in https://github.com/cal-itp/benefits/issues/1527. 

The code changes are associated with:

- GitHub workflows
- Azure DevOps/Terraform pipeline
- Python package versioning using `git` metadata

The last step needed to close #1527, will be to remove the `test` and `prod` branches and remove references to them in all the GitHub workflows. 

## Acceptance Criteria

- [ ] Merging to `main` deploys to the `dev` environment
- [ ] Pushing a pre-release tag deploys to the `test` environment
- [ ] Pushing a release tag deploys to the `prod` environment and triggers the release job
- [ ] The conditional `environment`, `dev` and `prod` Azure jobs (`terraform` stage) run correctly
- [ ] Python package versioning uses `git` metadata
- [ ] There is a guarantee that the required checks have passed on the commit to be deployed

### CODE REVIEW

File: .github/workflows/release.yml
Comment (thekaveman) [inline .github/workflows/release.yml:1]: I'd like this workflow to be steps in the `deploy` workflow, or else called specifically by the `deploy` workflow (rather than `on.push` etc.). 

Otherwise, we get the GitHub release published well before the code is actually deployed, which feels a little out of order.

**Review by thekaveman — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml
Comment (lalver1) [reply .github/workflows/release.yml:1]: Thanks! I agree, the `release` job should run after the `deploy` job is finished, so I'll add some steps in `deploy.yml` to achieve this. Oh and `release` should only run on an actual release, not a release candidate so I'll be careful about keeping this behavior unchanged.

**Comment by lalver1** [comment]
Now `deploy` and `release` are two sequential jobs in the `Deploy` workflow. I grouped the two `release` steps into the `release` job to set the `if` condition and the `permissions` just once at the job level instead of twice if they had been at the step level. 

Unfortunately I couldn't set `${{ github.ref_type != 'tag' && github.ref_name || contains(github.ref, '-rc') && 'test' || 'prod' }}` as an environment variable using `env` in the workflow and avoid duplicating writing it out since the `env` [context is not available](https://docs.github.com/en/enterprise-cloud@latest/actions/learn-github-actions/contexts#context-availability) at the `job` level, only at the `step` level.

With this implementation we'll have the following behavior in `main`:

- push a commit:
  - set `environment = main`
  - push image to GHCR
  - deploy to https://dev-benefits.calitp.org/
- push an rc tag:
  - set `environment = test`
  - push image to GHCR
  - deploy to https://test-benefits.calitp.org/
- push a tag:
  - set `environment = prod`
  - push image to GHCR
  - deploy to https://benefits.calitp.org/
  - if `deploy` succeeded and `deployment environment == prod` , make a GitHub release

**Review by thekaveman — CHANGES_REQUESTED** [review]
This is looking really good so far! I think the workflow updates all look correct.

But we still need the updates to the Azure DevOps/Terraform pipeline, which is looking at the old `test` and `prod` branches etc. 

Take a look at Server for how those pipelines work with tag-based.

File: .github/workflows/deploy.yml
Comment (thekaveman) [inline .github/workflows/deploy.yml:82]: I get that you're just re-using the same expression here, which is fine... but really all this check cares about is if it's a production tag, we don't need the env name or anything else here.

Maybe this could be simplified to something like `${{ github.ref_type == "tag" && !contains(github.ref, "-rc") }}`?

Comment (lalver1) [reply .github/workflows/deploy.yml:82]: That's true, this expression works better! 👍

**Review by thekaveman — DISMISSED** [review]
Also, don't forget that tag-based deployment means we get the version number from the tag instead of hardcoding it in the source.

That means changes at least to [`pyproject.toml`](https://github.com/cal-itp/benefits/blob/main/pyproject.toml#L3)

See this PR for how it was done in Server:

- https://github.com/cal-itp/eligibility-server/pull/382

**Comment by thekaveman** [comment]
@lalver1 you should be able to rebase on `main` now to get the Cypress tests passing.

Fixed by #2214

**Comment by lalver1** [comment]
Thanks @thekaveman, trying it now!

File: .github/workflows/deploy.yml
Comment (thekaveman) [inline .github/workflows/deploy.yml:82]: I think these have to be single quotes :roll_eyes:

**Comment by lalver1** [comment]
Thanks for the review, I had overlooked the Azure DevOps/Terraform pipeline and the dynamic versioning part. I think we should now have all the parts ready for tag-based deployment.

The only thing I don't feel very confident about is if we need `COPY .git .git`in the [dev container's Dockerfile](https://github.com/cal-itp/benefits/pull/2213/commits/48046721c0eb9beaff628e8e1eec2a5af2b69ee6#diff-13bd9d7a30bf46656bc81f1ad5b908a627f9247be3f7d76df862b0578b534fc6R10) since we already have this line in the app container's Dockerfile. If indeed we don't need it, I guess including it has no impact in how it works, it would just make the code cleaner.

File: terraform/pipeline/deploy.yml
Comment (thekaveman) [inline terraform/pipeline/deploy.yml:53]: This is correct to remove the old branch names here, but we still want to skip the `plan` under the same conditions as before. Namely, when we are doing a `test` or `prod` deployment. 

I think you can use a similar conditional as below in the `apply`.

**Review by thekaveman — CHANGES_REQUESTED** [review]
Getting really close @lalver1, this is looking great.

Just a few more small notes and I think it's ready.

**Comment by thekaveman** [comment]
> The only thing I don't feel very confident about is if we need COPY .git .gitin the [dev container's Dockerfile](https://github.com/cal-itp/benefits/pull/2213/commits/48046721c0eb9beaff628e8e1eec2a5af2b69ee6#diff-13bd9d7a30bf46656bc81f1ad5b908a627f9247be3f7d76df862b0578b534fc6R10) since we already have this line in the app container's Dockerfile. If indeed we don't need it, I guess including it has no impact in how it works, it would just make the code cleaner.

I was wondering the same thing. I see we do it this way in Server, so there must have been a reason. No worries, it is copied over immediately once we launch the devcontainer, so I think you can just leave it as-is.

File: .devcontainer/Dockerfile
Comment (thekaveman) [inline .devcontainer/Dockerfile:10]: On second thought (and after some testing), let's remove this -- I removed it locally and tried rebuilding the devcontainer after the appcontainer, and got an error:

```log
> [dev 8/8] RUN git init . &&     pre-commit install-hooks &&     rm -rf .git:
0.101 error: could not lock config file /calitp/app/.git/config: Permission denied
0.101 fatal: could not set 'core.repositoryformatversion' to '0'
```

This is a problem with the last few lines of this file, which should also be removed -- since the `.git` directory now exists, we can't call `git init` and do the temporary pre-commit install.

```
# install pre-commit environments in throwaway Git repository
# https://stackoverflow.com/a/68758943
COPY .pre-commit-config.yaml .
RUN git init . && \
    pre-commit install-hooks && \
    rm -rf .git
```

Even trying to install the hooks without `git init` fails due to some Docker layers weirdness... in any case, this matches how we have it in Server.

**Review by thekaveman — CHANGES_REQUESTED** [review]

File: .devcontainer/Dockerfile
Comment (thekaveman) [reply .devcontainer/Dockerfile:10]: In fact, take a look at what calitp.org does in `postAttach`, I think we should apply this here and in Server: https://github.com/cal-itp/calitp.org/blob/main/.devcontainer/postAttach.sh

File: .github/workflows/deploy.yml
Comment (lalver1) [reply .github/workflows/deploy.yml:82]: Ah that's true, thanks for catching that!

File: .devcontainer/Dockerfile
Comment (lalver1) [reply .devcontainer/Dockerfile]: Yep, I verified the behavior you described and made the suggested changes. The dev container loaded without a problem so we can make this change to eligibility-server as you mentioned.

File: terraform/pipeline/deploy.yml
Comment (lalver1) [reply terraform/pipeline/deploy.yml]: Yep, that makes sense. I used `parameters.tag_type` to determine if it is a `test` or `prod` deployment similar to the `apply` step and updated the condition. The condition may look a bit odd (maybe someone would expect keeping the `or` and just changing the `eq`s to `ne`s) but as far as I can tell the `and` is required for this step to run only for the cases when `apply` will not run. So this condition ensures that `# the plan is done as part of the apply (below), so don't bother doing it twice` following the comment for this `task`.

**Review by thekaveman — APPROVED** [review]
Nice work :+1: I'd like @angela-tran to take a look at this before you merge.

**Comment by lalver1** [comment]
Thanks @thekaveman! Yep, I won't merge until @angela-tran  has had a chance to look at it.

**Review by angela-tran — DISMISSED** [review]
This looks good to me! 

The only thing (which I don't think needs to block this PR) is in the GitHub deploy workflow, we aren't running the tests before deploying. Is this being done some other way?

**Review by thekaveman — CHANGES_REQUESTED** [review]
@angela-tran 

> The only thing (which I don't think needs to block this PR) is in the GitHub deploy workflow, we aren't running the tests before deploying. Is this being done some other way?

Good call, let's get this in as part of this PR.

Can we run the migrations and messages check as well?

@lalver1 see https://github.com/cal-itp/eligibility-server/blob/main/.github/workflows/docker-publish.yml#L24

File: terraform/azure-pipelines.yml
Comment (thekaveman) [reply terraform/azure-pipelines.yml:33]: I think it's probably because this line didn't change, in that the `workspace` is still set as an output variable by `workspace.py`: https://github.com/cal-itp/benefits/pull/2213/files/f4304d09dd1bf08a2797f165ba7d603cd0743a48#diff-33a8fda4d235c86352a2629618f4d0829b626df6084edef709ad22d97e8f80fcR52

File: terraform/azure-pipelines.yml
Comment (thekaveman) [inline terraform/azure-pipelines.yml:36]: These lines could be normalized to move the `TAG_TYPE` and `echo` in `tag.py` similar to how it works in `workspace.py`, but not a priority for this PR.

Comment (lalver1) [reply terraform/azure-pipelines.yml:36]: I agree. Great, I'll create an issue to come back to this later.

**Comment by lalver1** [comment]
Thanks @angela-tran, that's a good point about the tests! I'm thinking about reusing these workflows:

- `tests-ui`
- `tests-pytest`
- `tests-cypress`
- `check-migrations-and-messages`

I don't think we need to run `codeql` and `check-api` 🤔. If this makes sense, I'll add `on: workflow_call` to those workflows that don't already have it, and will call them from `deploy.yml` using `uses:` before the `deploy` job.

Oh and one more question, I'm thinking about setting `if: github.ref_type == 'tag'` when calling these tests, since we already run them as part of the branch protection status checks when `ref_type` is `branch` and we probably don't want to run them again?

**Comment by angela-tran** [comment]
> Thanks @angela-tran, that's a good point about the tests! I'm thinking about reusing these workflows:
> 
>   * `tests-ui`
>   * `tests-pytest`
>   * `tests-cypress`
>   * `check-migrations-and-messages`
> 
> 
> I don't think we need to run `codeql` and `check-api` 🤔. If this makes sense, I'll add `on: workflow_call` to those workflows that don't already have it, and will call them from `deploy.yml` using `uses:` before the `deploy` job.

That list looks good to me. And we'll want `deploy` to depend on the tests succeeding.

 
> Oh and one more question, I'm thinking about setting `if: github.ref_type == 'tag'` when calling these tests, since we already run them as part of the branch protection status checks when `ref_type` is `branch` and we probably don't want to run them again?

Yep, that's definitely needed. [`eligibility-server` does it too](https://github.com/cal-itp/eligibility-server/blob/main/.github/workflows/docker-publish.yml#L26) in its GitHub workflow.

**Comment by lalver1** [comment]
@angela-tran @thekaveman I added the tests to the `deploy` workflow and [required them to successfully complete](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) before the `deploy` job starts.

I did notice that `eligibility-server` has these [lines](https://github.com/cal-itp/eligibility-server/blob/main/.github/workflows/docker-publish.yml#L30-L32) in its `docker-publish` workflow
```
deploy:
    needs: test
    # !cancelled() is needed because if the whole workflow was cancelled, we don't want this job to run.
    # (github.ref_type != 'tag' || needs.test.result == 'success') is needed because if `test` did run, we only want this to run if `test` succeeded.
    if: (!cancelled() && (github.ref_type != 'tag' || needs.test.result == 'success'))
```

and I'm wondering if I should add `if: (!cancelled()` to the benefits `deploy` job since I guess we could manually want to cancel a deploy before it finishes running. I think the second part `&& (github.ref_type != 'tag' || needs.test.result == 'success')` may not be needed because 

```
needs:
      [tests-ui, tests-pytest, tests-cypress, check-migrations-and-messages]
```
covers this case?

**Comment by angela-tran** [comment]
> I'm wondering if I should add `if: (!cancelled()` to the benefits `deploy` job since I guess we could manually want to cancel a deploy before it finishes running. I think the second part `&& (github.ref_type != 'tag' || needs.test.result == 'success')` may not be needed because
> 
> ```
> needs:
>       [tests-ui, tests-pytest, tests-cypress, check-migrations-and-messages]
> ```
> 
> covers this case?

Great catch @lalver1 -- yes, I think we should add `if: (!cancelled())`. 

I took a closer look at that second part from `eligibility-server`, and I think you're right that it's not needed. GitHub docs on [`needs`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) confirms that the jobs listed must complete successfully before the dependent job will run, so it would be redundant to include this condition. I think it's a remnant from an [older version](https://github.com/cal-itp/eligibility-server/pull/329#discussion_r1341766075) of the expression.

Long story short 😄, we don't need the second part here, and it could also be removed in `eligibility-server`.

**Comment by lalver1** [comment]
Thanks @angela-tran! I added `if: (!cancelled())` to the `deploy` job and will make a ticket to remove the second part of the expression in `eligibility-server`; I think this one is ready for a review 🙂

**Review by thekaveman — APPROVED** [review]
I think this is ready to go :tada:

**Pull request merged**
